# Basic R

When learning a new programming language, it is customary to start with an "hello world" example. As simple as it is, this one line of code will ensure that we know how to print a string in output and how to execute code within cells in a notebook.
- Note: To execute the R code in the code cell below, click on the cell to select it and press Shift + Enter.

In [1]:
# Try your first R output

print('Hello, R!')

[1] "Hello, R!"


After executing the cell above, you should see that R prints Hello, R!. Congratulations on running your first R code!

### What version of R are we using?

Which version of R is running on this computer? We can ask directly R and obtain a detailed answer. Try executing the following code.
- Note: R.version displays many system-specific information, including operating system (os) and the R version in use.

In [2]:
R.version

               _                           
platform       x86_64-pc-linux-gnu         
arch           x86_64                      
os             linux-gnu                   
system         x86_64, linux-gnu           
status                                     
major          3                           
minor          3.2                         
year           2016                        
month          10                          
day            31                          
svn rev        71607                       
language       R                           
version.string R version 3.3.2 (2016-10-31)
nickname       Sincere Pumpkin Patch       

### Writing comments in R

In addition to writing code, note that it is always a good idea to add comments to your code. It will help others understand what you were trying to accomplish (the reason why you wrote a given snippet of code). Not only does this help other people understand your code, it can also serve as a reminder to you when you come back to it weeks or months later.

To write comments in R, use the number symbol # before writing your comment. When you run your code, R will ignore everything past the # on a given line.

In [3]:
# Practice on writing comments

print('Hello, R!') # This line prints a string
# print('Hi')
# Comments are great for writing notes alongside your code!

[1] "Hello, R!"


After executing the cell above, you should notice that This line prints a string did not appear in the output, because it was a comment (and thus ignored by R).

The second line was also not executed because 'print('Hi')' was preceded by the number sign (#) as well! Since this isn't an explanatory comment from the programmer, but an actual line of code, we might say that the programmer commented out that second line of code.

### Errors in R

Everyone makes mistakes. For many types of mistakes, R will tell you that you have made a mistake by giving you an error message. It is important to read error messages carefully to really understand where you made a mistake and how you may go about correcting it.

For example, if you spell print as frint, R will display an error message. 

In [ ]:
# Print string as error message

frint("Hello, R!")

The error message tells you:

- where the error occurred (more useful in large notebook cells or scripts), and
- what kind of error it was (NameError)

Here, R attempted to run the function frint, but could not determine what frint is since it's not a built-in function and it has not been previously defined by us either.

You'll notice that if we make a different type of mistake, by forgetting to close the string, we'll obtain a different error (i.e., a SyntaxError). 

In [4]:
# Try to see build in error message

print("Hello, R!)

ERROR: Error in parse(text = x, srcfile = src): <text>:3:7: unexpected INCOMPLETE_STRING
2: 
3: print("Hello, R!)
         ^


### Does R know about your error before it runs your code?

R is what is called an interpreted language. Compiled languages examine your entire program at compile time, and are able to warn you about a whole class of errors prior to execution. In contrast, R interprets your script line by line as it executes it. R will stop executing the entire program when it encounters an error (unless the error is expected and handled by the programmer, a more advanced subject that we'll cover later on in this course).

Try to run the code in the cell below and see what happens.

In [5]:
# Print string and error to see the running order

print("This will be printed")
frint("This will cause an error")
print("This will NOT be printed")

[1] "This will be printed"


ERROR: Error in eval(expr, envir, enclos): could not find function "frint"


# Problem 1

Generations of programmers have started their coding careers by simply printing "Hello, world!". You will be following in their footsteps.

#### Question 1

In the code cell below, use the print() function to print out the phrase, "Hello, world!".

In [6]:
#your code here


#### Question 2

Now, let's enhance your code with a comment. In the code cell below, print out the phrase, "Hello, world!". This time, add a comment next to your code that says, "comments are great for adding notes alongside your code".
- Note: Remember the # is used to add comments in a code cell.

In [9]:
#your code here


# Problem 2

Problem 2 covers the different data types in R. 

Everything in R is an object.

There are many different types of objects in R. R has 6 basic data types.

- character
- numeric (real or decimal)
- integer
- logical
- complex

There are some examples:

- character: "a", "swc"
- numeric: 2, 15.5
- integer: 2L (the L tells R to store this as an integer)
- logical: TRUE, FALSE
- complex: 1+4i (complex numbers with real and imaginary parts)

R provides many functions to examine features of vectors and other objects, for example

- class() - what kind of object is it (high-level)?
- typeof() - what is the object’s data type (low-level)?
- length() - how long is it? What about two dimensional objects?
- attributes() - does it have any metadata?

The following code cells contain some examples.

#### Intructions

You can get R to tell you the type of an expression by using the built-in typeof() function.

In [16]:
x <- "dataset"

typeof(x)

[1] "character"

In [17]:
x <- 5

typeof(x)

[1] "double"

In [18]:
x <- 5L

typeof(x)

[1] "integer"

In [20]:
x <- "My name is ..."

typeof(x)

[1] "character"

#### Question 1

In the code cell below, use the typeof() function to check the object type of 12.0.

In [22]:
#your code here


### Integers

Here are some examples of integers. Integers can be negative or positive numbers:

<a align="center">
    <img src="https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/PY0101EN/Chapter%201/Images/TypesInt.png" width="600">
</a>

We can verify this is the case by using, you guessed it, the typeo() function:

In [26]:
# Print the type of -1

typeof(-1L)
typeof(1)

[1] "integer"

[1] "double"

In [27]:
# Print the type of 4

typeof(4L)

[1] "integer"

In [35]:
# Print the type of 0

typeof(0)

[1] "double"

### Double 

Doubles represent real numbers; they are a superset of integer numbers but also include "numbers with decimals".

Once again, you can test some examples with the typeof() function:

In [29]:
# Print the type of 1.0

typeof(1.0)

[1] "double"

In [30]:
# Print the type of 0.5

typeof(0.5)

[1] "double"

#### Question 2

Print the type of 1.56 using the typeof() function. 

In [33]:
#your code here


# Problem 3

Problem 3 deals with converting from one object type to a different object type.

#### Instructions 

You can change the type of the object in R; this is called typecasting. For example, you can convert an integer into a character (e.g. 2 to "2").

In [36]:
# Verify that this is an character

typeof(as.character(2))

[1] "character"

#### Converting integers to double

Let us now cast integer 2 to double.

In [41]:
# Convert 2 to a float
as.double(2L)

[1] 2

In [42]:
# Convert integer 2 to a double and check its type

typeof(as.double(2))

[1] "double"

When we convert an integer into a double, we don't really change the value (i.e., the significand) of the number. However, if we cast a double into an integer, we could potentially lose some information. For example, if we cast the double 1.1 to integer we will get 1 and lose the decimal information (i.e., 0.1):

In [43]:
# Casting 1.1 to integer will result in loss of information

as.integer(1.1)

[1] 1

#### Converting from strings to integers or double

Sometimes, we can have a string that contains a number within it. If this is the case, we can cast that string that represents a number into an integer using as.integer():

In [44]:
# Convert a string into an integer

as.integer('1')

[1] 1

But if you try to do so with a string that is not a perfect match for a number, you'll get a warning message. R produces an NA values - **not available**. Try the following:

In [45]:
# Convert a string into an integer with error
as.integer('1 or 2 people')

Warning message in eval(expr, envir, enclos):
"NAs introduced by coercion"

[1] NA

You can also convert strings containing decimal point numbers into double objects.
- Note: Strings can be represented with single quotes ('1.2') or double quotes ("1.2"), but you can't mix both (e.g., "1.2').

In [47]:
# Convert the string "1.2" into a float

as.double('1.2')

[1] 1.2

#### Converting numbers to strings

If we can convert characters to numbers, it is only natural to assume that we can convert numbers to characters, right?

In [48]:
# Convert an integer to a string

as.character(1)

[1] "1"

There is no reason why we shouldn't be able to make doubles into strings as well.



In [49]:
# Convert a float to a string

as.character(1.2)

[1] "1.2"

### Logical

Logical is another important type in R. An object of type Boolean can take on one of two values: TRUE or FALSE:

In [50]:
# Value true

TRUE

[1] TRUE

Notice that the value TRUE is in uppercase. The same is true for FALSE (i.e. you must use the uppercase FALSE).

In [51]:
# Value false

FALSE

[1] FALSE

When you ask R to display the type of a logical object it will show logical which stands for logical:

In [52]:
# Type of TRUE

typeof(TRUE)

[1] "logical"

In [53]:
# Type of FALSE

typeof(FALSE)

[1] "logical"

We can cast logical objects to other data types. If we cast a boolean with a value of TRUE to an integer or float we will get a one. If we cast a boolean with a value of FALSE to an integer or double we will get a zero. Similarly, if we cast a 1 to a Logical, you get a True. And if we cast a 0 to a Logical we will get a FALSE. Let's give it a try:

In [54]:
# Convert TRUE to integer

as.integer(TRUE)

# Convert FALSE to integer
as.integer(FALSE)

[1] 1

[1] 0

#### Question 1

What is the data type of the result of: 6 / 2?

In [56]:
#your code here


#### Question 2

Convert the string '20' into a double. Verify your answer with the typeof() function.

In [59]:
#your code here


# Problem 4

Problem 4 deals with expressions and variables. 

Expressions in R can include operations among compatible types (e.g., integers and doubles).

#### Instructions

Here is an example of a basic arithmetic operation of adding multiple numbers.

In [64]:
# Addition operation expression

43 + 60 + 16 + 41

[1] 160

We can perform subtraction operations using the minus operator. In this case the result is a negative number.

In [65]:
# Subtraction operation expression

50 - 60

[1] -10

We can do multiplication using an asterisk.

In [66]:
# Multiplication operation expression

5 * 5

[1] 25

We can also perform division with the forward slash:

In [67]:
# Division operation expression

25 / 6

[1] 4.166667

#### Question 1

Write an expression that calculates how many hours there are in 160 minutes.

In [68]:
#your code here


#### Instructions

R follows well accepted mathematical conventions when evaluating mathematical expressions. In the following example, R adds 30 to the result of the multiplication (i.e., 120).

In [70]:
# Mathematical expression

30 + 2 * 60

[1] 150

Just like mathematics, expressions enclosed in parentheses have priority. So the following multiplies 32 by 60.

In [71]:
# Mathematical expression

(30 + 2) * 60

[1] 1920

# Problem 5

Problem 5 has to do with storing values in variables, so we can use them later on. 

#### Instructions 

Below we are going to store the addition of values 43, 60, 16, and 41, into x.

In [73]:
# Store value into variable

x = 43 + 60 + 16 + 41

To see the value of x in a Notebook, we can simply place it on the last line of a cell.

In [74]:
# Print out the value in variable

x

[1] 160

We can also perform operations on x and save the result to a new variable.

In [75]:
# Use another variable to store the result of the operation between variable and value

y = x / 60
y

[1] 2.666667

If we save a value to an existing variable, the new value will overwrite the previous value:

In [76]:
# Overwrite variable with new value

x = x / 60
x

[1] 2.666667

It's a good practice to use meaningful variable names, so you and others can read the code and understand it more easily.

In [77]:
# Name the variables meaningfully

total_min = 43 + 42 + 57 # Total length of albums in minutes
total_min

[1] 142

In [78]:
# Name the variables meaningfully

total_hours = total_min / 60 # Total length of albums in hours 
total_hours

[1] 2.366667

In the cells above we added the length of three albums in minutes and stored it in total_min. We then divided it by 60 to calculate total length total_hours in hours. You can also do it all at once in a single expression, as long as you use parenthesis to add the albums length before you divide, as shown below.

In [79]:
# Complicate expression

total_hours = (43 + 42 + 57) / 60  # Total hours in a single expression
total_hours

[1] 2.366667

If you'd rather have total hours as an integer, you can of course replace the floating point division with integer division (i.e., //).

#### Question 1

What is the value of x where x = 3 + 2 * 2

In [ ]:
#your code here


#### Question 2

What is the value of y where y = (3 + 2) * 2?

In [ ]:
#your code here


#### Question 3

What is the value of z where z = x + y?

In [ ]:
#your code here
